In [19]:
import random
import time

#1000 elementos desordenados aleatoriamente entre 1 y 1000
arreglo_desordenado = random.sample(range(1, 10000001), 10000000)

In [20]:

import math


def merge(left_arr, right_arr):
    merged = []
    i, j = 0, 0

    left_arr.append(math.inf)
    right_arr.append(math.inf)

    while left_arr[i] != math.inf or right_arr[j] != math.inf:
        if left_arr[i] < right_arr[j]:
            merged.append(left_arr[i])
            i += 1
        else:
            merged.append(right_arr[j])
            j += 1

    left_arr.pop()
    right_arr.pop()

    return merged


def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


In [21]:
start_time = time.time()
merge_sort(arreglo_desordenado)
end_time = time.time()

print("Tiempo de ejecución del merge sort normal:", end_time - start_time, "segundos")


Tiempo de ejecución del merge sort normal: 83.45572876930237 segundos


Utilizando arreglos aleatorios de 1,000,000 elementos, se encontraron los siguientes tiempos: 

Tiempo de ejecución del merge sort paralelo: 8.491748332977295 segundos (En un script por problemas de windows)

Tiempo de ejecución del merge sort normal: 4.509132385253906 segundos

El Normal casi es el doble de rápido que al correr procesos en paralelo, por lo que se puede decir que el costo de la paralelización y su unión posterior es mucho más grande que los beneficios que trae. 


Sin embargo, al utilizar arreglos aleatorios de 10,000,000 elementos, se encontraron los siguientes tiempos:

Tiempo de ejecución del merge sort normal: 83.45572876930237 segundos

Tiempo de ejecución del merge sort paralelo: 103.79995441436768 segundos

Aqui vemos como los costos se van acercando a un punto de quiebre donde eventualmente podría atraer mayor beneficio si se utilizaran más datos.
